src/hooks/useResizer.js

```diff
@@ -0,0 +1,124 @@
+import { useEffect, useRef } from "react";
+import PropTypes from "prop-types";
+
+import { addActions, actions } from "../actions";
+import { defaultState } from "./useTableState";
+
+defaultState.resizedColumns = {};
+
+addActions({
+  resizeColumn: "__resizeColumn__"
+});
+
+const propTypes = {
+  columns: PropTypes.arrayOf(
+    PropTypes.shape({
+      width: PropTypes.number
+    })
+  ),
+  onResizedChange: PropTypes.func
+};
+
+export const useResizer = props => {
+  PropTypes.checkPropTypes(propTypes, props, "property", "useResizer");
+
+  const {
+    columns,
+    hooks: { headers: headerHooks },
+    state: [{ resizedColumns }, setState],
+    getResizerProps,
+    onResizedChange
+  } = props;
+  const currentlyResizingInfo = useRef(undefined);
+
+  useEffect(() => {
+    if (resizedColumns) {
+      Object.keys(resizedColumns).forEach(index => (columns[index].width = resizedColumns[index]));
+    }
+  }, [resizedColumns]);
+
+  const addResizer = (columns, api) => {
+    let bottomLevelColumnCount = 0;
+
+    columns.forEach(column => {
+      if (!column.columns) {
+        const index = bottomLevelColumnCount;
+
+        column.isBottomLevel = true;
+
+        column.resizable = column.resizable !== undefined ? column.resizable : true;
+
+        column.getResizerProps = () => ({
+          ...(column.resizable && {
+            onMouseDown: e => onDragStart(e, column, index),
+
+            // prop overrides
+            ...(getResizerProps && getResizerProps)
+          })
+        });
+
+        bottomLevelColumnCount++;
+      }
+    });
+
+    return columns;
+  };
+
+  headerHooks.push(addResizer);
+
+  const resizeColumn = e => {
+    const { clientX: currentPosition } = e;
+    const { index, initialPosition, initialWidth } = currentlyResizingInfo.current;
+    const positionXDelta = currentPosition - initialPosition;
+    const minWidth = columns[index].minWidth || 10;
+    const maxWidth = columns[index].maxWidth || null;
+
+    let updatedWidth = initialWidth + positionXDelta > 0 ? initialWidth + positionXDelta : 0;
+
+    if (updatedWidth < minWidth) {
+      updatedWidth = minWidth;
+    }
+
+    if (maxWidth && updatedWidth > maxWidth) {
+      updatedWidth = maxWidth;
+    }
+
+    setState(old => {
+      let newResizedColumns = { ...old.resizedColumns };
+
+      newResizedColumns[index] = updatedWidth;
+
+      if (onResizedChange) {
+        onResizedChange(resizedColumns, e);
+      }
```

Comment:
We don't use this callback style anymore. If users want to detect changes to internal state, they are supposed to hoist the state out (preferably using `useTableState`) and watch it in their own code using an effect or something.

Reply:
Fair enough.  That's actually better since there's less bulk features that not all users will tap into.